Kernel/Arch/i686/Bootloader/stivale2.c

```diff
@@ -193,26 +194,26 @@ static void stage1MapBootloader(void)
 __attribute__((section(".early_text")))
 static void stage1MapHighMemory(void)
 {
-    uint32_t kernelDirectoryEntryIdx = _KERNEL_START >> PAGE_DIR_ENTRY_SHIFT;
+    uint32_t kernelDirectoryEntryIdx = (uint32_t)&_KERNEL_START >> PAGE_DIR_ENTRY_SHIFT;
 
     // First kernel page table entry
     struct DirectoryEntry* kernelDirectoryEntry = &pageDirectory.entries[kernelDirectoryEntryIdx];
     kernelDirectoryEntry->present = 1;
     kernelDirectoryEntry->readWrite = 1;
-    kernelDirectoryEntry->tableAddr = (uint32_t)&kernelPageTable;
+    kernelDirectoryEntry->tableAddr = (uint32_t)&kernelPageTable >> PAGE_TABLE_ENTRY_SHIFT;
 
     // Second kernel page table entry
     struct DirectoryEntry* pagesDirectoryEntry = kernelDirectoryEntry + 1;
     pagesDirectoryEntry->present = 1;
     pagesDirectoryEntry->readWrite = 1;
-    pagesDirectoryEntry->tableAddr = (uint32_t)&pagesPageTable;
+    pagesDirectoryEntry->tableAddr = (uint32_t)&pagesPageTable >> PAGE_TABLE_ENTRY_SHIFT;
 
-    size_t kernelMemoryIdx = 0;
     for (uintptr_t addr = KERNEL_START; addr < KERNEL_END; addr += ARCH_PAGE_SIZE) {
-        struct TableEntry* kernelMemoryTableEntry = &lowMemoryPageTable.pages[kernelMemoryIdx++];
+        size_t kernelMemoryIdx = addr >> PAGE_TABLE_ENTRY_SHIFT;
+        struct TableEntry* kernelMemoryTableEntry = &kernelPageTable.pages[kernelMemoryIdx & 0xfff];
         kernelMemoryTableEntry->present = 1;
         kernelMemoryTableEntry->readWrite = 1;
-        kernelMemoryTableEntry->frame = addr;
+        kernelMemoryTableEntry->frame = kernelMemoryIdx - 0xc0000;
```

Comment:
Should probably make a `#define` for the `0xC000`. Is there some way we could get that from the linker script?

Reply:
Noting that `(KERNEL_BASE >> PAGE_TABLE_ENTRY_SHIFT)` is the correct replacement for `0xC0000`.